pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go

```diff
@@ -85,22 +85,65 @@ func init() {
 					(*scpb.IndexName)(nil),
 					(*scpb.IndexPartitioning)(nil),
 					(*scpb.IndexComment)(nil),
+					(*scpb.IndexColumn)(nil),
 				),
 				joinOnIndexID(from.el, to.el, "table-id", "index-id"),
 				targetStatusEq(from.target, to.target, scpb.ToPublic),
 				currentStatus(from.node, scpb.Status_BACKFILL_ONLY),
 				currentStatus(to.node, scpb.Status_PUBLIC),
 			}
 		})
+	registerDepRule(
+		"index existence precedes index dependents",
+		scgraph.SameStagePrecedence,
+		"index", "index-column",
+		func(from, to nodeVars) rel.Clauses {
+			return rel.Clauses{
+				from.el.Type(
+					(*scpb.PrimaryIndex)(nil),
+					(*scpb.SecondaryIndex)(nil),
+				),
+				to.el.Type(
+					(*scpb.IndexColumn)(nil),
+				),
+				joinOnIndexID(from.el, to.el, "table-id", "index-id"),
+				targetStatusEq(from.target, to.target, scpb.ToPublic),
+				currentStatus(from.node, scpb.Status_BACKFILL_ONLY),
+				currentStatus(to.node, scpb.Status_PUBLIC),
+			}
+		})
+	registerDepRule(
+		"temp index existence precedes index dependents",
+		scgraph.SameStagePrecedence,
+		"index", "index-column",
+		func(from, to nodeVars) rel.Clauses {
+			return rel.Clauses{
+				from.el.Type(
+					(*scpb.PrimaryIndex)(nil),
+					(*scpb.SecondaryIndex)(nil),
+				),
+				to.el.Type(
+					(*scpb.IndexColumn)(nil),
+				),
+				joinOnIndexID(from.el, to.el, "table-id", "index-id"),
+				targetStatus(from.target, scpb.Transient),
+				targetStatus(to.target, scpb.ToPublic),
+				currentStatus(from.node, scpb.Status_DELETE_ONLY),
+				currentStatus(to.node, scpb.Status_PUBLIC),
+			}
+		})
 
 	registerDepRule(
-		"partitioning set right after temp index existence",
+		"partitioning and columns set right after temp index existence",
```

Comment:
set aside IndexParitioning, how is this rule different from the above one where we state "temporary index must exist right before its dependents (really just its index columns) becomes public"?